- https://www.calculatorsoup.com/calculators/discretemathematics/combinations.php

- https://jovian.com/ashutoshgole18/module-8-inferential-statistics

In [1]:
import random
n = random.randint(1, 20)
r = random.randint(1, n) # 0 < r <= n

n, r

(10, 3)

In [2]:
def fact(n):
  if n <= 1:
    return 1
  else:
    return n * fact(n-1)

fact(n), fact(r)

(3628800, 6)

In [3]:
def per(n, r):
  return int(fact(n) / fact(n-r))

per(n, r)

720

In [4]:
def com(n, r):
  return int(per(n, r) / fact(r))

com(n, r)

120

In [5]:
def compute_lcm(x, y):
   if x > y:
       greater = x
   else:
       greater = y

   while(True):
       if((greater % x == 0) and (greater % y == 0)):
           lcm = greater
           break
       greater += 1

   return lcm

compute_lcm(4, 3)

12

In [6]:
def ratio(e, s):
  lcm = compute_lcm(e, s) 
  hcf = (s*e) / lcm
  return f'{int(e/hcf)} / {int(s/hcf)}'

ratio(8, 52)

'2 / 13'

In [7]:
def dec_rat(n):
  s = str(n).split('.')
  a = int(s[0] + s[1])
  b = len(s[1])
  c = 10 ** b
  return ratio(a, c)

In [8]:
from fractions import Fraction
y = Fraction(2, 4)
y

Fraction(1, 2)

# Important results

In [9]:
com(n, r = 0)

1

In [10]:
com(n, n)

1

In [11]:
a = com(n, r)
b = com(n, n-r)

a, b, a == b

(120, 120, True)

In [12]:
a = r * com(n, r)
b = n * (com(n-1, r-1))

a, b, a == b

(360, 360, True)

In [13]:
a = com(n, r) + com(n, r-1)
b = com(n+1, r)

a, b, a == b

(165, 165, True)

In [14]:
com(n, 0) == com(n, n) == 1

True

In [15]:
com(n, 1) == com(n, n-1) == n

True

In [16]:
a = com(n, r)
b = per(n, r) / fact(r)

a, b, a == b

(120, 120.0, True)

Permutations
-------

Permutations: Find the number of ways in which the letters of the word ‘MOSAIC’ can be rearranged to form different six-letter words.

    30

    100

    720

    320

----------

    720

    ✓ Correct
    Feedback:

The word `MOSAIC` contains `six` letters with which we need to form six-letter words. 

Therefore, the formula that you would be using is `n!`. 

Since n = 6 here, from the permutation formula, we have 

    n! = 6*5*4*3*2*1 = 720.

In [17]:
# Ans.
fact(6)

720

In [18]:
per(6, 6)

720

Permutations
-------

Again, using the letters of the word ‘MOSAIC’, can you find out how many three-letter words can be formed?

    100

    300

    130

    120

--------------

    120

    ✓ Correct
    Feedback:

The word '`MOSAIC`' contains `six` letters with which we need to form `three-letter` words. 

Therefore, the formula that you would be using is n!/(n-r)!. 

Since n = 6 and r = 3 here, from the permutation formula, we have 

    n!/(n-r)! = 6!/3! = 720/6 = 120.

In [19]:
6*5*4

120

In [20]:
per(6, 3)

120

In [21]:
# Ans.
fact(6) / fact(6-3)

120.0

Combinations
--------

In the case of combinations, you need not take the order into account while finding the number of ways to arrange a group of objects.

------------------

Combinations: 

Suppose you have a list of 10 customers who have given a bad rating to your product. 

Now, you need to randomly pick three customers so that you can interview them and take their feedback. 

In how many ways can you choose three customers from the given set of 10?

    120

    100

    300

    400

---------

    120

    ✓ Correct
    Feedback:

The formula for combinations is n!/(r! (n-r)!). 

In the given problem, n = 10 and r = 3. 

Substituting the value in the formula, you get 

    (10)!/3! 7! = 120.

In [22]:
# Ans.
com(10, 3)

120

Combinations
-------

In a group of 6 boys and 4 girls, four children are to be selected. 

In how many different ways can they be selected?

    210

    100

    110

    140

-----------

    210

    ✓ Correct
    Feedback:

Yes that's the correct answer, there are a total of 10 children (6+4) and we want to select four of them, therefore this is a combination problem.
 
    10!/ [4! * (10-4)!] = 210

In [23]:
# Ans.
com(6+4, 4)

210

Sample space: 
-----------

A sample space is nothing but the list of all possible outcomes of a random experiment. 

It is denoted by S = {all the possible outcomes}. 

For example, in the coin toss example, the sample space is S = {H, T}, where H = heads and T = tails.

Event: 
----------

It is a subset, i.e., a part of the sample space that you want to be true for your probability experiment. 

For example, if in a coin toss you want heads to be the desired outcome, then the event becomes {H}. 

As you can see clearly, {H} is a part of {H, T}.

Concepts of probability
-----------

Let's say you have a box containing five cards with the letters 'B', 'R', 'E', 'A' and 'D' written on them. 

Now, you are asked to pick a card randomly from the box, and each time you pick a card, you note down the letter written on it and keep the card outside the box. 

You continued this process five times until no cards were left in the box. 

What is the probability that you picked the letters in the same sequence as the word 'BREAD'?

    1/24

    3/24

    1/120

    1/90

------------

    1/120

    ✓ Correct
    Feedback:

For calculating the probability, you first need to find the number of outcomes in the sample space, 

i.e., the number of ways in which you can pick five letters from the box. 

This is a case of permutation since the order is of importance for forming a word. 

Therefore, the total number of outcomes in the sample space is 

    5! = 120

The event here would be the outcome where you pick the cards in the exact sequence 'BREAD'. 

Hence, the total number of outcomes in the event would be 1. 

Since, probability = (Total number of outcomes in the event) / (Total number of outcomes in sample space), this value comes out to be 1/120.


In [24]:
s = fact(5)
s # sample space

120

In [25]:
e = 1
e # event

1

In [26]:
lcm = compute_lcm(e, s) 
lcm

120

In [27]:
hcf = (s*e) / lcm
hcf

1.0

In [28]:
f'{e/hcf} / {s/hcf}'

'1.0 / 120.0'

In [29]:
ratio(e, s)

'1 / 120'

Properties of probability
-----------

Which of the following options cannot be the probability of a particular event in any scenario?

    -0.2

    1.4

    0.1

    Both -0.2 and 1.4

----------

    Both -0.2 and 1.4

    ✓ Correct
    Feedback:

Recall the rules of probability. 

It states that the probability of any event must lie between 0 and 1. 

Both -0.2 and 1.4 do not lie in this range. 

Hence, this is the correct option.

Probability
-----------

Two dice are thrown simultaneously. 

What is the probability that they will display two numbers whose product is even?

    1/2

    1/9

    7/16

    3/4

-----------

    3/4

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

The total number of favourable outcomes is 27: 

{(1, 2), (1, 4), (1, 6), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (3, 2), (3, 4), (3, 6), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (5, 2), (5, 4), (5, 6), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), and (6, 6)}. 

    So the probability becomes = 27/36 = 3/4

In [30]:
s = 6*6
s

36

In [31]:
e = 6*3 + 3*3
e # (6 any and 3 even) or (3 even and 3 even)

27

In [32]:
lcm = compute_lcm(e, s) 
lcm

108

In [33]:
hcf = (s*e) / lcm
hcf

9.0

In [34]:
f'{e/hcf} / {s/hcf}'

'3.0 / 4.0'

In [35]:
ratio(e, s)

'3 / 4'

Probability
-------------

Mr. Lee has two children; the first one is a boy. 

What is the probability that the second child is a girl?

    1/2

    3/4

    1/3

    4/9

--------

    1/2

    ✓ Correct
    Feedback:

Yes that’s the right answer, the total number of favourable outcomes is one, and the total number of outcomes is two. 

    Thus, the probability P = 1/2.


Properties of Probability
----------

Assume that for a particular experiment there are five possible outcomes. 

Now the probabilities of the first three outcomes are given as 0.1, 0.2 and 0.3, respectively. 

You don't know the exact probabilities of the last two outcomes, but you do know that the probability of the fifth outcome is three times the probability of the fourth outcome. 

Can you use the rules of probability to find the probabilities of the fourth and fifth outcomes?

    0.1 and 0.3

    0.12 and 0.4

    0.2 and 0.6

    0.3 and 0.9

------------

    0.1 and 0.3

    ✓ Correct
    Feedback:

Take the probability of the fourth outcome as x. 

Thus, as per the question, the probability of the fifth outcome would be 3x. 

Now, the total probability of all the outcomes 

    = 0.1+ 0.2 +0.3 + x +3x = 0.6 + 4x. 

Now from the rules of probability, we have that the sum of all the probabilities for a particular experiment turns out to be 1. 

Thus, we can write, 0.6 + 4x = 1, solving which we get x = 0.1. 

Thus, the probability of the fourth outcome is 0.1 and that of the fifth outcome is 0.3.

In [36]:
# 0.1 + 0.2 + 0.3 + x + 3*x = 1
# 4*x = 1 - 0.6 = 0.4
# x = 4 / 40 = 1 / 10

1/10, 3/10

(0.1, 0.3)

In [37]:
from sympy.solvers import solve
from sympy import Symbol

x = Symbol('x')
x

x

In [38]:
# ?solve
y = solve(0.1 + 0.2 + 0.3 + x + 3*x - 1)
y

[0.100000000000000]

In [39]:
y = round(y[0], 1)
y, 3*y

(0.1, 0.3)

Types of events
-------------

Consider the following statements and choose the correct option.

`Statement 1:` 

Event A denotes the scenario where the marketing campaign leads to more than 500 subscribers, whereas event B denotes the scenario where the marketing campaign leads to less than 200 subscribers.

`Statement 2:` 

Event A denotes the scenario where the marketing campaign leads to less than 500 subscribers, whereas event B denotes the scenario where the marketing campaign gets more than 500 subscribers.

    In statement 1, the events are disjoint, whereas, 
    in statement 2, the events are independent. 

    In statement 1, the events are independent, whereas, 
    in statement 2, the events are disjoint. 

    Both statements denote events which are disjoint. 

    Both statements denote events which are independent. 

-----------

    Both statements denote events which are disjoint. 

    ✓ Correct
    Feedback:

Two events are considered independent when one event has no effect on the other. 

Two events are considered disjoint if both of them cannot occur simultaneously.

If you go through statement 1, you can clearly see that the events are disjoint, since the subscribers cannot be more than 500 and less than 200 at the same time. 

Similarly, in the second statement, the events are again disjoint, since the number of subscribers cannot be both more than 500 and less than 500 at the same time.


Complement Rule of Probability
----------

Let A denote the event 'The campaign leads to more than 500 subscribers'. 

What would be its complementary event?

    The campaign would lead to less than 200 subscribers.

    The campaign would lead to more than 200 subscribers.

    The campaign would lead to less than or equal to 500 subscribers.

    The campaign would lead to more than or equal to 500 subscribers.

------------

    The campaign would lead to less than or equal to 500 subscribers.

    ✓ Correct
    Feedback:

The complement of event A is essentially the negation of the given scenario. 

A denotes the situation where the number of subscribers is more than 500 and hence A' would denote the situation where the number of subscribers is not more than 500, 

    i.e., it is less than or equal to 500.

Probability Values
-----------

Let's say Box8 has planned another similar marketing campaign called 'XYZ' in the next quarter in Bengaluru. 

Now the analyst has found out that the probability of the current marketing campaign "ABC" to get more than 500 subscribers is 0.9. 

Is it possible to calculate the probability that the "XYZ" campaign in Bengaluru would also get more than 500 subscribers based on this value of 0.9?

    No

    Yes

-------------

    No

    ✓ Correct
    Feedback:

Observe that both the events in question are independent in nature. 

Hence, you can't predict one's probability based on the other event. 

    Therefore the answer is No.

Addition rule
------------

A and B are two candidates applying for the job role of an analyst at a leading analytics firm in Delhi. 

They have cleared all the previous rounds and are now set for the final interview after which only one of them will be selected for the job. 

Candidate A has a probability of 0.3 of getting the job and candidate B has a probability of 0.25 of getting the job. 

Now, what is the probability that one of the two gets the job?

    0.40

    0.45

    0.55

    0.75

-------------

    0.55

    ✓ Correct
    Feedback:

Here P(A) = 0.3 and P(B) = 0.25. 

You need to find P(A⋃B). 

Also, observe that P(A) and P(B) are disjoint, i.e., both of them cannot get the job simultaneously. 

Now, P(A⋃B) = P(A)+P(B)-P(A⋂B). 

Since, the events are disjoint, P(A⋂B) = 0.

Probability
----------

A few cards are numbered from 1 to 20 and are mixed up. 

One card is then drawn at random. 

What is the probability that this card has a number that is a multiple of four or nine?

    1/5

    7/20

    3/5

    9/20



---

    7/20

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

Addition rule is to be applied, P(A⋃B) = P(A)+P(B) − P(A⋂B) where

    A - Event where the number is divisible by 4
    B -  Event where the number is divisible by 9

Since there are no numbers between 1 to 20 that are divisible by both 4 and 9 

    P(A⋂B) becomes 0.

P(A) = 5/20 and P(B) = 2/20

So the probability, 

    P(A⋃B) = 5/20 + 2/20 = 7/20.

Probability
--------

If a card is chosen from a standard deck of cards, what is the probability of it being a five or a seven?

    1/13

    2/13

    7/16

    9/11

----------

    2/13

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

Addition rule is to be applied, P(A⋃B) = P(A)+P(B) − P(A⋂B) where

    A - Event where the card contains the number five
    B -  Event where the card contains the number seven

Since there are no card that contains numbers both five and seven,

    P(A⋂B) becomes 0.

P(A) = 4/52 and P(B) = 4/52

So the probability, 

    P(A⋃B) = 4/52 + 4/52 = 8/52 = 2/13


In [40]:
e, s = 4 + 4, 52
ratio(e, s)

'2 / 13'

Addition vs Multiplication
-----------

Consider the following two statements and choose the correct option.


`Statement 1:` 

The probability of Rahul playing in today's match is 0.25, whereas that of Bumrah is 0.35. 

You need to find the probability that either of them will play in today's match.


`Statement 2:` 

The probability of Rahul playing in today's match is 0.25, whereas that of Bumrah is 0.35. 

You need to find the probability that both of them will play in today's match.

    Statement 1 uses the multiplication rule, 
    whereas statement 2 uses the addition rule to 
    find the probability that is being asked.

    Statement 1 uses the addition rule, 
    whereas statement 2 uses the multiplication rule to 
    find the probability that is being asked.

    Both of them use the addition rule to 
    find the probability that is being asked.

    Both of them use the multiplication rule to 
    find the probability that is being asked.

----------

    Statement 1 uses the addition rule, whereas 
    statement 2 uses the multiplication rule to 
    find the probability that is being asked.

    ✓ Correct
    Feedback:

Observe that in the first statement `either` of the events can occur. 

So, there is an 'OR' involved here. 

That is, you need to find P('Rahul' OR 'Bumrah'). 

In the second statement, however, both the events are occurring simultaneously.

Hence, there is an 'AND' involved here, i.e., you need to find P('Rahul' AND 'Bumrah'). 

Therefore, in the first statement, you will be using the addition rule to find the probability, 

whereas in the second statement, you will use the multiplication rule.

Addition and Multiplication Rules
---------

Now, what is the probability that either Rahul or Bumrah will play in today's match, given that their individual probabilities of playing are 0.25 and 0.35, respectively? 

Also, it is mentioned that both the events are not dependent on each other. 

    0.69

    0.51

    0.56

    0.009

---------

    0.51

    ✓ Correct
    Feedback:

Let 'A' denote the event that Rahul plays in the match and 'B' denote the event that Bumrah plays in the match. 

It is given that P(A) = 0.25 and P(B) = 0.35. 

    Also, it is given that A and B are independent. 

You need to find P(A⋃B). 

Now, this is given by the formula P(A⋃B) = P(A) + P(B) - P(A⋂B). 

    Now,
    P(A) = 0.25
    P(B) = 0.35

P(A⋂B) = P(A and B) = P(A)*P(B) 

= 0.25*0.35 = 0.0875 

(since A and B are independent)

Therefore, the final probability is 

    0.25 + 0.35 - 0.0875 = 0.5125 ~ 0.51.

    Probability of event A or B

P(A U B) = P(A) + P(B) - P(A ⋂ B)

    for dependent event,

P(A ⋂ B) = 0

    for independent event,
    
P(A ⋂ B) = P(A) * P(B)

In [41]:
0.25 + 0.35 - 0.25 * 0.35

0.5125

Addition and Multiplication Rules
-----------

A bag contains six balls: three black ones and three red ones. 

You take out a ball from the bag, note its colour and put it back in the bag. 

You then draw another ball from the bag and note down its colour. 

What is the probability that balls of the same colour were drawn twice?

    1/2

    1/4

    1/6

    9/10

----------------

    1/2

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

We have to use both addition and multiplication rule,

P(A⋃B) = P(A) + P(B) − P(A⋂B) where

    A - Event where both the balls drawn are black balls
    B -  Event where both the balls drawn are red balls


In this case we can observe that events are disjoint 

    and hence P(A⋂B) = 0

Now to calculate  event A  we use multiplication rule, 

as both the balls have to be black, 

    so P(A) = P(black ball) * P(black ball) 

P(A) = 3/6 * 3/6 = 1/4

    and P(B) = P(red ball) * P(red ball) 

P(B) = 3/6 * 3/6 = 1/4

So the probability, P(A⋃B) 

    = 1/4  + 1/4 = 2/4 = 1/2

In [42]:
0.5 * 0.5 + 0.5 * 0.5 - 0

0.5

In [43]:
ratio(5, 10)

'1 / 2'

# Practice Questions

Permutations and Combinations
------

Given below are two statements

`Statement 1:` 

You need to choose five bowlers from a given roster of seven bowlers for the next cricket match.

`Statement 2:` 

You want to find the number of ways in which you can arrange your seven family members on seven seats for a family photograph.

Choose the correct option:

    The first statement uses permutation, 
    whereas the second statement uses combination.

    Both the statements use permutation.

    The first statement uses combination, 
    whereas the second statement uses permutation.

    Both the statements use combinations.

------------


    The first statement uses combination, 
    whereas the second statement uses permutation.

    ✓ Correct
    Feedback:

The first statement asks you to choose something, whereas the second statement asks you to order something.

Hence, the first statement uses combination, whereas the second statement uses permutation.

Permutations and Combinations
--------

You have a list of seven letters, of which three are vowels (A, I and E) and four are consonants (B, C, D and F). 

You need to use these letters to make four-letter words in such a way that they should contain exactly `two vowels` from the given three `and` exactly `two consonants` from the given four. 

How many such words can be formed?

    576

    432

    459

    600

------

    432

    ✓ Correct
    Feedback:

You can choose two vowels from the given three vowels in 

    3!/(2!*1!) = 3 ways. 

Similarly, you can choose two consonants from the given four consonants in 

    4!/(2!2!) = 6 ways. 

Since these two events are independent, you have 3*6 = 18 combinations of four unique letters. 

Now, you can arrange each of the 18 combinations in 4! ways. 

Thus, the total number of words that can be formed are 

    18*4! = 18*24 = 432.

In [44]:
com(3, 2) * com(4, 2) * fact(4)

432

Properties of Probability
-----------

In a random experiment, there were a total of three outcomes. 

Now, the probability of the second and third outcomes came out to be twice that of the first outcome. 

Therefore, the sum of the probabilities of the first and second outcomes comes out to be ____.

    0.4

    0.6

    0.9

    0.76

--------

    0.6

    ✓ Correct
    Feedback:

Let the first outcome have a probability of x. 

Therefore, the second and third outcomes have a probability of 2x each. 

Using the total sum rule of probability, we have 

    x + 2x + 2x = 1

    5x = 1

    x = 0.2

Therefore, the sum of the first and second probabilities comes out to be 

    0.2 + 0.4 = 0.6

In [45]:
# x + 2*x + 2*x = 1

y = solve(x + 2*x + 2*x - 1)
y

[1/5]

In [46]:
x + 2*x

3*x

In [47]:
z = y[0] + 2*y[0]
z

3/5

In [48]:
type(z)

sympy.core.numbers.Rational

In [49]:
float(z)

0.6

Rules of Probability
-------------

A college has been offering two finance programs, AGFM and BGFM, to its students for the last five years. 

From past historical data, it has been found that the probability of a student joining the AGFM program is 0.4 and the probability of a student joining the BGFM program is 0.475. 

Now, if a student joins one program, he/she cannot obviously enrol for the next program. 

So, what is the probability that a student enrols in either of the programs?

    0.625

    0.875

    0.568

    0.889

-------------

    0.875

    ✓ Correct
    Feedback:

Note that we are calculating P(AGFM or BGFM). 

Hence, we are going to use the addition rule of probability. 

Now, it is given that 

    P(AGFM) = 0.4 and P(BGFM) = 0.475

Also, it has been mentioned that both the enrolment events are `disjoint`.

    Therefore, P(AGFM and BGFM) = 0. 

Therefore, by the addition rule, we have 

    P(AGFM or BGFM) = 0.4 + 0.475 = 0.875

Probability
--------

A jar contains four marbles: three red ones and one white one. 

Two marbles are drawn but are replaced after each draw. 

What is the probability that marbles of the same colour are drawn twice?

    5/8

    1/2

    1/3

    4/5

-------------

    5/8

    ✓ Correct
    Feedback:

Yes that’s the right answer. Addition rule is to be applied, 

    P(A⋃B) = P(A)+P(B) − P(A⋂B) where

    A - Event where both the marbles drawn are red marbles
    B -  Event where both the marbles drawn are white marbles

In this case we can observe that events are disjoint 

    and hence P(A⋂B) = 0

Now to calculate event A we use the multiplication rule, 

as both the marbles have to be red , 

    so P(A) = P(red marble) * P(red marble) 

P(A) = 3/4 * 3/4 = 9/16

    and P(B) = P(white marble) * P(white marble) 

P(B) = 1/4 * 1/4 = 1/16

So the probability, 

    P(A⋃B) = 9/16  + 1/16 = 5/8.

In [55]:
z = Fraction(3/4 * 3/4) + Fraction(1/4 * 1/4)
str(z)

'5/8'

In [54]:
dec_rat(3/4 * 3/4 + 1/4 * 1/4)

'5 / 8'

Probability
-----------

A number is selected at random from the first thirty natural numbers. 

What is the chance that it is a multiple of either three or thirteen?

    2/5

    1/4

    6/7

    11/18

-------

    2/5

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

Addition rule is to be applied, 

    P(A⋃B) = P(A)+P(B) − P(A⋂B) where

    A - Event where the number is divisible by 3
    B -  Event where the number is divisible by 13

Since there are no numbers between 1 to 30 that are divisible by both 3 and 13 

    P(A⋂B) becomes 0.

P(A) = 10/30 and 

P(B) = 2/30

So the probability, 

    P(A⋃B) = 10/30 + 2/30 = 2/5.

In [63]:
e = 30//3
s = 30//13

e, s

(10, 2)

In [64]:
ratio(e + s, 30)

'2 / 5'

Probability
------------

In a throw of dice, what is the probability of getting a number greater than five?

    5/6

    1/3

    1/6

    1/2

------------

    1/6

    ✓ Correct
    Feedback:

Yes that's the correct answer. 

The number of favourable outcomes is one. 

As six is the only number greater than five, 

and the total number of outcomes are six (1,2,3,4,5,6) 

    so the probability becomes 1/6.

In [66]:
ratio(1, 6)

'1 / 6'

Probability
------------

A bag contains `seven red` balls and `eleven green` balls. 

If a ball is drawn out at random from it, what is the probability that it is red?

    7/18

    4/11

    3/7

    11/18

------

    7/18

    ✓ Correct
    Feedback:

Yes that’s the right answer. 

The total number of favourable outcomes is the total number of red balls which is 7 and 

the total number of outcomes is 18 (7 red + 11 green balls). 

    Thus, P = 7/18.

In [70]:
ratio(7, 7 + 11)

'7 / 18'